### PR TITLE
Standardize duration measurement to microseconds

### PR DESCRIPTION
**Pull request by delano**
## Overview

This PR standardizes duration measurement across Familia to use microseconds consistently. Previously, the codebase mixed microsecond precision (for accurate measurement) with millisecond display units, creating confusion in logs and instrumentation hooks. Now all duration values are measured and reported in microseconds, simplifying the mental model and making performance analysis more precise.

## Key Changes

**Duration Measurement Consolidation**
- Added `Familia.now_in_μs` utility method (with `now_in_microseconds` alias) that wraps `Process.clock_gettime(Process::CLOCK_MONOTONIC, :microsecond)`
- Replaced direct `Process.clock_gettime` calls throughout the codebase with the new utility
- Eliminated redundant millisecond conversion calculations (`/ 1000.0`)

**API Simplification**
- Instrumentation hooks now receive `duration` in microseconds instead of `duration_ms` in milliseconds
- Log messages now show `duration=1234` (microseconds) instead of `duration_ms=1.23` (milliseconds)
- DatabaseLogger middleware simplified by removing conversion overhead

**Affected Components**
- `Familia::Horreum` initialization and persistence lifecycle
- `Familia::Instrumentation` API (command, pipeline, and lifecycle hooks)
- `DatabaseLogger` middleware structured logging output
- Documentation and examples updated throughout

## Migration Notes

**For instrumentation hook consumers:**
```ruby
# Before (v2.x)
Familia.on_command do |cmd, duration_ms, ctx|
  StatsD.timing("familia.#{cmd}", duration_ms)  # Already in milliseconds
end

# After (v3.0)
Familia.on_command do |cmd, duration, ctx|
  StatsD.timing("familia.#{cmd}", duration / 1000.0)  # Convert to milliseconds
end
```

The change makes the API more honest - you get the raw microsecond measurement and can convert to whatever unit your monitoring system expects.

## Testing

Added comprehensive test coverage for the new utility methods and verified edge cases:
- Microsecond utilities return correct types and valid measurements
- Middleware command tracking works with empty and populated arrays
- CommandMessage objects maintain required fields
- Duration measurements accurately capture sleep intervals

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

___

### **PR Type**
Enhancement

___

### **Description**
- Add `Familia.now_in_μs` utility method for consistent microsecond duration measurement

- Standardize all duration measurements to use integer microseconds instead of float milliseconds

- Update instrumentation API: rename `duration_ms` parameter to `duration` across all hooks

- Simplify logging and middleware by removing millisecond conversion calculations

- Add comprehensive test coverage for new time utility methods and edge cases

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Process.clock_gettime calls"] -->|"Replace with"| B["Familia.now_in_μs utility"]
  C["duration_ms parameter<br/>Float milliseconds"] -->|"Rename to"| D["duration parameter<br/>Integer microseconds"]
  E["Instrumentation hooks"] -->|"Updated to use"| D
  F["DatabaseLogger middleware"] -->|"Simplified with"| B
  G["Horreum lifecycle<br/>initialize/save"] -->|"Use new utility"| B
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>utils.rb</strong><dd><code>Add microsecond utility methods for duration measurement</code>&nbsp; </dd></summary>
<hr>

lib/familia/utils.rb

<ul><li>Add <code>now_in_μs</code> method that wraps <br><code>Process.clock_gettime(Process::CLOCK_MONOTONIC, :microsecond)</code><br> <li> Add <code>now_in_microseconds</code> alias for better discoverability<br> <li> Returns Integer microseconds for consistent duration measurement</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-c0dff2491f5f69a56bfb313f500f326639dc868f6a9ed0f271667ef04edb6e73">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>horreum.rb</strong><dd><code>Standardize Horreum initialization to use microseconds</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/horreum.rb

<ul><li>Replace <code>Process.clock_gettime</code> calls with <code>Familia.now_in_μs</code> in <br><code>initialize</code> method<br> <li> Remove millisecond conversion arithmetic (<code>/ 1000.0).round(2)</code><br> <li> Rename <code>duration_ms</code> parameter to <code>duration</code> in lifecycle logging and <br>instrumentation<br> <li> Store durations as integer microseconds</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-c42ea61e6354fd846082c12370a61f8e7771e39f1779388fad96d72658207300">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>persistence.rb</strong><dd><code>Standardize Horreum persistence to use microseconds</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/horreum/persistence.rb

<ul><li>Replace <code>Process.clock_gettime</code> calls with <code>Familia.now_in_μs</code> in <code>save</code> <br>method<br> <li> Remove millisecond conversion calculations<br> <li> Rename <code>duration_ms</code> to <code>duration</code> in debug logging and instrumentation <br>hooks<br> <li> Store durations as integer microseconds throughout persistence <br>lifecycle</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-6e8aa91a474a9506b89e49a01bd5a0ca61383a0b45b319adc266a0008fb8d6d2">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>instrumentation.rb</strong><dd><code>Update instrumentation API for microsecond durations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/instrumentation.rb

<ul><li>Update <code>on_command</code> hook documentation: rename <code>duration_ms</code> to <code>duration</code>, <br>change type to Integer microseconds<br> <li> Update <code>on_pipeline</code> hook documentation: rename <code>duration_ms</code> to <code>duration</code>, <br>change type to Integer microseconds<br> <li> Update <code>on_lifecycle</code> hook documentation: rename <code>duration_ms</code> to <code>duration</code> <br>in context hash<br> <li> Update <code>notify_command</code> and <code>notify_pipeline</code> methods to use <code>duration</code> <br>parameter<br> <li> Add examples showing conversion to milliseconds when needed (e.g., <br><code>duration / 1000.0</code>)</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-64e27bffc57883f4002f2ee38cbf7def7248a5e6147b3903f37ffa2907f28dcc">+15/-15</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>database_logger.rb</strong><dd><code>Simplify DatabaseLogger middleware for microseconds</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/middleware/database_logger.rb

<ul><li>Replace local <code>Process.clock_gettime</code> call with <code>Familia.now_in_μs</code> in <br><code>now_in_μs</code> method<br> <li> Remove millisecond conversion calculations from structured logging <br>output<br> <li> Rename <code>duration_ms</code> to <code>duration</code> in log messages and instrumentation <br>hook calls<br> <li> Simplify logging output to show only <code>duration=<microseconds></code> instead of dual <br><code>duration_μs</code> and <code>duration_ms</code><br> <li> Update <code>call</code>, <code>call_pipelined</code>, and <code>call_once</code> methods to pass integer <br>microseconds to instrumentation hooks</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-f03d37a9450eee7bffdd92769de8493c57eec4739b36640062e12c7a36805ed3">+8/-13</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>logging.rb</strong><dd><code>Update logging documentation for microseconds</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/familia/logging.rb

<ul><li>Update debug method documentation example: change <code>duration_ms: 1.23</code> to <br><code>duration: 1234</code><br> <li> Update info method documentation example: change <code>duration_ms: 2.34</code> to <br><code>duration: 2340</code><br> <li> Update log output format to show integer microseconds instead of float <br>milliseconds</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-0c27ddafd0cd0db6b531e9484273e3599146d250054d3487f58c0064e7d62fd1">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>middleware_sampling_try.rb</strong><dd><code>Add middleware command tracking test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

try/unit/core/middleware_sampling_try.rb

<ul><li>Add test for <code>clear_commands</code> with empty command arrays<br> <li> Add test for <code>clear_commands</code> idempotency<br> <li> Add test verifying CommandMessage objects have required fields <br>(command, μs, timeline)<br> <li> Add test ensuring commands array contains no nil elements after <br>operations</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-9c10b8e139234f10ebe5228c384cbe365b9ada9cd850628f0f13763def713fec">+29/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils_try.rb</strong><dd><code>Add comprehensive time utility test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

try/unit/core/utils_try.rb

<ul><li>Add test for <code>Familia.now</code> returning Float timestamp in seconds<br> <li> Add test for <code>Familia.now</code> accepting Time argument<br> <li> Add test for <code>Familia.now_in_μs</code> returning Integer microseconds<br> <li> Add test for <code>now_in_microseconds</code> alias functionality<br> <li> Add test for monotonic clock behavior with sleep interval measurement</ul>

</details>

  </td>
  <td><a href="https://github.com/delano/familia/pull/167/files#diff-0809265f202586eecf61b2cf957c66c61235dd4f8dcd231142adcac2323e9dea">+33/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: try/unit/core/middleware_sampling_try.rb [hunk 1]
```diff
@@ -195,7 +195,7 @@
 end
 
 # 4 pipeline operations captured (CommandMessage uses Data.define)
-pipeline_count = commands.select { |cmd| cmd.command.include?(' | ') }.count
+pipeline_count = commands.compact.select { |cmd| cmd.command.include?(' | ') }.count
```
[the remark this answers is not in the document]
Comment (delano) [reply try/unit/core/middleware_sampling_try.rb]: The tests pass without this defensive change  for the teardown in this one file. We're add thready safety coverage in another branch now to hopefully smoke this issue out of its hole.
